src/mongo_types.ts

```diff
@@ -440,3 +442,61 @@ export class TypedEventEmitter<Events extends EventsDescription> extends EventEm
 
 /** @public */
 export class CancellationToken extends TypedEventEmitter<{ cancel(): void }> {}
+
+/**
+ * Helper types for dot-notation filter attributes
+ */
+
+/** @public */
+export type Join<T extends unknown[], D extends string> = T extends []
+  ? ''
+  : T extends [string | number]
+  ? `${T[0]}`
+  : T extends [string | number, ...infer R]
+  ? `${T[0]}${D}${Join<R, D>}`
+  : string;
+
+/** @public */
+export type PropertyType<Type, Property extends string> = string extends Property
+  ? unknown
+  : Property extends keyof Type
+  ? Type[Property]
+  : Property extends `${number}`
+  ? Type extends ReadonlyArray<infer ArrayType>
+    ? ArrayType
+    : unknown
+  : Property extends `${infer Key}.${infer Rest}`
+  ? Key extends `${number}`
+    ? Type extends ReadonlyArray<infer ArrayType>
+      ? PropertyType<ArrayType, Rest>
+      : unknown
+    : Key extends keyof Type
+    ? Type[Key] extends Map<string, infer MapType>
+      ? MapType
+      : PropertyType<Type[Key], Rest>
+    : unknown
+  : unknown;
+
+// We dont't support nested circular references
```

Comment:
You could add support for circular types using an accumulator as depth checker and bail out when too deep. This works around the TS error "TS2589: Type instantiation is excessively deep and possibly infinite.". A max depth of 10 is probably enough for most cases without making the TS checker too slow.

Example

```typescript
type O<T, K extends string, Prefix extends string, Depth extends number[]> = K extends keyof T ? { [_ in `${Prefix}.${K}`]: T[K] } & (T[K] extends object ? SubObjects<T[K], Extract<keyof T[K], string>, `${Prefix}.${K}`, [...Depth, 1]> : {}) : {};

type SubObjects<T, K extends string, Prefix extends string, Depth extends number[]> =
    Depth['length'] extends 10 ? {} : //bail out when too deep
        K extends keyof T ? T[K] extends Array<infer A> ? SubObjects<A, Extract<keyof A, string>, `${Prefix}.${K}.${number}`, [...Depth, 1]> :
            T[K] extends object ? O<T[K], Extract<keyof T[K], string>, Prefix extends '' ? K : `${Prefix}.${K}`, Depth> : { [P in Prefix extends '' ? K :`${Prefix}.${K}`]: T[K] } : {};

type FilterQuery<T> = SubObjects<T, Extract<keyof T, string>, '', []>;
```

[Playground link](https://www.typescriptlang.org/play?#code/FASwdgLgpgTgZgQwMZQAQAUYHsAmBXJCVAb2FXNRBwC5Uw8BbAI1gG4yKIQIAbKWgM4QY4AObsKqLrygAmQcLESKWAO5hYtAKoC2wAL7BQkWIhSodsEh3JVa9Znsl5dMMAgb9UQkWHE3UAAdsfEIBWkxcAggAbQBdZXIGBHBI0IgIkOj2Q2AIAE9AtAB5AB4AFQAaVABpVCgAD2gwHAFvRT9qzCg4EAb6pqgWtp8xaoARKECIAAsB5ta6RhYYeIA+VABeWvmhxYBrKHysOFRy1AB+ElQYgH1KMFQAAwASYm7ehv0AOjea-SecVo5RiNTiqH0qAAZKgABQgsG7YaoLBMABWUEIl1QAGU8ExiujMRABBVQXFqgBRJowZAQUqHY6nBEU9q+URraqvd4wHp9H5-AHVGLfUWTaYzaoARjiG1oxH0AEpUPL9Ow8oU0HiCUSwhVqnVGgsRh1RF1eZ8kYtRp1UOLZla2g4VustgF7TMYgByPh+WZe8FGvZtKUABmxCpVqAA9NGmCkeCi8ERVDMhlIsFhUDgoFMApJDYNkYyTmdsSzHagAIIwWn5UrgOBWKsbK7awkYvVVqk0ukMo6l7tssSc55vD7837Ef5T52wIU3UXfD3S2Uq-OSCgVoPI1GdohXMosnvCPsl5nk6o2jnmvn9HeLL1e7F1WjcidfKf-J4TKazOXXDE6APBgFp9JWT4vlG75gZ+gqAsC5IQlGCrqgURSoAAYiAPDQDAACKeCwPW5QbNs7a6iS+qoNSp6EP2TJnFepqjk+wqymhmpSFKWxYTheGEcRpSWDAazsEgWBgEIKK0BAPHbKQkgGKwQA)

<img width="1438" alt="Screenshot 2022-01-16 at 06 00 16" src="https://user-images.githubusercontent.com/450980/149647946-d1b1cfbd-1c09-4550-a273-79d4c6f4e8a8.png">

Reply:
Thanks for the suggestion @marcj! We're going to look into this further (I created [NODE-3875](https://jira.mongodb.org/browse/NODE-3875) for tracking). For now we have released 4.3.1 which contains a fix for this issue in the case of directly self-referential types.